File: src/main/scala/com/vc/hard/L996.scala
package com.vc.hard

object L996 {
  def numSquarefulPerms(arr: Array[Int]): Int = {
    import scala.collection.mutable

    val countMap = new mutable.HashMap[Int, Int]()
    val squareMap = new mutable.HashMap[Int, mutable.HashSet[Int]]()

    arr.indices.foreach(i => {
      countMap.put(arr(i), countMap.getOrElse(arr(i), 0) + 1)
    })

    val n = arr.length
    var res = 0
    arr.indices.foreach(i => {
      (i + 1 until n).foreach(j => {
        val x = arr(i)
        val y = arr(j)
        val c = Math.sqrt(x + y)
        if(c == Math.floor(c)) {
          val setY = squareMap.getOrElse(x, mutable.HashSet[Int]())
          setY += y
          squareMap.put(x, setY)

          val setX = squareMap.getOrElse(y, mutable.HashSet[Int]())
          setX += x
          squareMap.put(y, setX)
        }
      })
    })

    def solve(num: Int, remaining: Int): Unit = {
      if(remaining == 0) res += 1
      else {
        countMap.put(num, countMap(num) - 1)
        if(squareMap.contains(num)) {
          squareMap(num).foreach(s => {
            if(countMap.getOrElse(s, 0) > 0) {
              solve(s, remaining - 1)
            }
          })
        }
        countMap.put(num, countMap(num) + 1)
      }
    }

    countMap.foreach(x => {
      solve(x._1, arr.length - 1)
    })
    res
  }
}

File: src/main/java/com/vc/hard/L996.java
package com.vc.hard;

import java.util.*;

class L996 {
    int count = 0;
    public int numSquarefulPerms(int[] arr) {
        HashMap<Integer, HashSet<Integer>> map = new HashMap<>();

        for(int i = 0; i < arr.length; i++) {
            for(int j = i + 1; j < arr.length; j++) {
                int from = arr[i];
                int to = arr[j];

                double sqrt = Math.sqrt(from + to);
                if(sqrt == Math.ceil(sqrt)) {
                    HashSet<Integer> fromList = map.getOrDefault(from, new HashSet<Integer>());
                    fromList.add(to);

                    HashSet<Integer> toList = map.getOrDefault(to, new HashSet<Integer>());
                    toList.add(from);

                    map.put(from, fromList);
                    map.put(to, toList);
                }
            }
        }

        //System.out.println(map);

        //{1=[8], 17=[8], 8=[1, 17]}
        HashMap<Integer, Integer> countMap = new HashMap<>();
        for(int i = 0; i < arr.length; i++) {
            countMap.put(arr[i], countMap.getOrDefault(arr[i], 0) + 1);
        }

        for(Integer from: countMap.keySet()) {
            solve(from, arr.length - 1, map, countMap);
        }
        return count;
    }

    private void solve(int from, int remaining,
                       HashMap<Integer, HashSet<Integer>> map, HashMap<Integer, Integer> countMap) {
        if(countMap.getOrDefault(from, 0) <= 0) return;

        if(remaining == 0) count++;

        //System.out.println(from);
        countMap.put(from, countMap.get(from) - 1);
        if(map.containsKey(from)) {
            for(Integer to: map.get(from)) {
                solve(to, remaining - 1, map, countMap);
            }
        }
        countMap.put(from, countMap.get(from) + 1);
    }
}
